### PR TITLE
[BEAM-886] Implement annotation based NewDoFn in Python SDK

### PR DESCRIPTION
**Pull request by sb2nov**
- Implement the new annotation based DoFn in python see https://s.apache.org/a-new-dofn 
- Migrate all the cases of DoFn to NewDoFn
- CallableWrapperDoFn now extends NewDoFn
- Typechecking is also handled using the NewDoFn
- All unittests pass locally

Future Work:
- Remove OldDoFn completely from the code
- Deprecate the use of Context and pass state directly to the process method

R: @robertwb PTAL

---

Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by robertwb** [comment]
To ease reviewing, could you split all the migrations of DoFn to NewDoFn to a separate commit?

**Comment by sb2nov** [comment]
@robertwb Done

File: sdks/python/apache_beam/runners/common.py [hunk 1]
```diff
@@ -71,99 +72,214 @@ def __init__(self,
...
+      self.state = state
+
+      # SideInputs
+      global_window = window.GlobalWindow()
+      self.side_inputs = [side_input
+                          if isinstance(side_input, sideinputs.SideInputMap)
+                          else {global_window: side_input}
+                          for side_input in side_inputs]
+
+      self.args = args if args else []
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/common.py]: Why is this needed now?

File: sdks/python/apache_beam/runners/common.py [hunk 2]
```diff
@@ -71,99 +72,214 @@ def __init__(self,
                # Preferred alternative to context
                # TODO(robertwb): Remove once all runners are updated.
                state=None):
-    self.has_windowed_side_inputs = False  # Set to True in one case below.
-    if not args and not kwargs:
+    # TODO(Sourabhbajaj): Remove the usage of OldDoFn
+    if isinstance(fn, core.NewDoFn):
+      self.is_new_dofn = True
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/common.py]: Have you been testing with Cython installed?

File: sdks/python/apache_beam/runners/common.py [hunk 3]
```diff
@@ -71,99 +72,214 @@ def __init__(self,
...
+    if self.has_windowed_side_inputs and len(element.windows) > 1:
+      for w in element.windows:
+        self.context.set_element(
+            WindowedValue(element.value, element.timestamp, (w,)))
+        self._process_outputs(element, self.dofn_process(self.context))
+    else:
+      self.context.set_element(element)
+      self._process_outputs(element, self.dofn_process(self.context))
+
+  def new_dofn_process(self, element):
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/common.py]: This code is performance critical, have you run any benchmarks? (Preferably of the compiled version, which is where it matters most as that's what runs on the workers.)

File: sdks/python/apache_beam/runners/common.py [hunk 4]
```diff
@@ -71,99 +72,214 @@ def __init__(self,
...
+    self.context.set_element(element)
+    arguments, _, _, defaults = self.dofn.get_function_arguments('process')
+    defaults = defaults if defaults else []
+
+    self_in_args = int(isinstance(self.dofn.process, types.MethodType) and \
+        self.dofn.process.im_self is not None and \
+        'self' in arguments)
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/common.py]: I'm a bit wary of relying on the name of the argument being "self" to branch here (despite being a common convention).

File: sdks/python/apache_beam/runners/common.py [hunk 5]
```diff
@@ -71,99 +72,214 @@ def __init__(self,
...
+
+    # Call for the process function for each window if has windowed side inputs
+    # otherwise we can optimize the runner by calling process for entire window
+    # set in one go.
+    if all(isinstance(s, dict) or s.is_globally_windowed()
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/common.py]: Nit: s -> si

File: sdks/python/apache_beam/runners/common.py [hunk 6]
```diff
@@ -71,99 +72,214 @@ def __init__(self,
...
+           for s in self.side_inputs):
+      windows = [(window.GlobalWindow(), element.windows)]
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/common.py]: If s is a dict, how do you know that it's the global window?

File: sdks/python/apache_beam/transforms/core.py [hunk 7]
```diff
@@ -228,10 +324,9 @@ def __init__(self, fn):
 
     self._fn = fn
     if _fn_takes_side_inputs(fn):
-      self.process = lambda context, *args, **kwargs: fn(
-          context.element, *args, **kwargs)
+      self.process = fn
     else:
-      self.process = lambda context: fn(context.element)
+      self.process = lambda e: fn(e)
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/core.py]: Both of these clauses can be "self.process = fn"

File: sdks/python/apache_beam/transforms/core.py [hunk 8]
```diff
@@ -577,7 +672,7 @@ class ParDo(PTransformWithSideInputs):
   def __init__(self, fn_or_label, *args, **kwargs):
     super(ParDo, self).__init__(fn_or_label, *args, **kwargs)
 
-    if not isinstance(self.fn, DoFn):
+    if not (isinstance(self.fn, DoFn) or isinstance(self.fn, NewDoFn)):
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/core.py]: not isinstance(self.fn, (DoFn, NewDoFn))

File: sdks/python/apache_beam/transforms/core.py [hunk 9]
```diff
@@ -588,7 +683,9 @@ def infer_output_type(self, input_type):
         self.fn.infer_output_type(input_type))
 
   def make_fn(self, fn):
-    return fn if isinstance(fn, DoFn) else CallableWrapperDoFn(fn)
+    if isinstance(fn, DoFn) or isinstance(fn, NewDoFn):
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/core.py]: same

File: sdks/python/apache_beam/transforms/core.py (hunk 8 above)
Comment (robertwb) [inline sdks/python/apache_beam/transforms/core.py]: Or should there be a common baseclass?

File: sdks/python/apache_beam/transforms/core.py (hunk 9 above)
Comment (sb2nov) [inline sdks/python/apache_beam/transforms/core.py]: Done

File: sdks/python/apache_beam/transforms/core.py (hunk 8 above)
Comment (sb2nov) [inline sdks/python/apache_beam/transforms/core.py]: Done. As we'll remove the oldDoFn soon I don't think it might be worthwhile to create an abstract class

File: sdks/python/apache_beam/transforms/core.py (hunk 7 above)
Comment (sb2nov) [inline sdks/python/apache_beam/transforms/core.py]: Done

File: sdks/python/apache_beam/runners/common.py (hunk 6 above)
Comment (sb2nov) [inline sdks/python/apache_beam/runners/common.py]: This is based on the existing logic in oldDoFn https://github.com/apache/beam/blob/python-sdk/sdks/python/apache_beam/runners/common.py#L85

File: sdks/python/apache_beam/runners/common.py (hunk 5 above)
Comment (sb2nov) [inline sdks/python/apache_beam/runners/common.py]: Fixed

File: sdks/python/apache_beam/runners/common.py (hunk 4 above)
Comment (sb2nov) [inline sdks/python/apache_beam/runners/common.py]: Fixed

File: sdks/python/apache_beam/runners/common.py (hunk 3 above)
Comment (sb2nov) [inline sdks/python/apache_beam/runners/common.py]: Running some benchmarks right now

File: sdks/python/apache_beam/runners/common.py (hunk 1 above)
Comment (sb2nov) [inline sdks/python/apache_beam/runners/common.py]: Just makes the followup code a bit easier

File: sdks/python/apache_beam/runners/common.py [hunk 10]
```diff
@@ -71,99 +72,214 @@ def __init__(self,
...
                # TODO(robertwb): Remove once all runners are updated.
                state=None):
-    self.has_windowed_side_inputs = False  # Set to True in one case below.
-    if not args and not kwargs:
+    # TODO(Sourabhbajaj): Remove the usage of OldDoFn
+    if isinstance(fn, core.NewDoFn):
+      self.is_new_dofn = True
+      self.step_name = step_name
+      self.window_fn = windowing.windowfn
+      self.tagged_receivers = tagged_receivers
```
Comment (sb2nov) [inline sdks/python/apache_beam/runners/common.py]: Done

File: sdks/python/apache_beam/runners/common.py (hunk 2 above)
Comment (sb2nov) [inline sdks/python/apache_beam/runners/common.py]: Thanks for catching, should be fixed now.

**Comment by sb2nov** [comment]
@robertwb PTAL again. Still running benchmarks so will post the results but the code changes you mentioned should be in.

File: sdks/python/apache_beam/transforms/core.py [hunk 11]
```diff
@@ -119,6 +120,112 @@ def aggregate_to(self, aggregator, input_value):
...
+  In order to have concrete DoFn objects one has to subclass from DoFn and
+  define the desired behavior (start_bundle/finish_bundle and process) or wrap a
+  callable object using the CallableWrapperDoFn class.
+  """
+
+  ElementParam = 'ElementParam'
+  ContextParam = 'ContextParam'
+  SideInputParam = 'SideInputParam'
+  TimestampParam = 'TimestampParam'
+  WindowsParam = 'WindowsParam'
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/core.py]: We should only be exposing one window at a time.

File: sdks/python/apache_beam/runners/common.py (hunk 3 above)
Comment (robertwb) [inline sdks/python/apache_beam/runners/common.py]: Any update on this?

File: sdks/python/apache_beam/transforms/core.py (hunk 11 above)
Comment (sb2nov) [inline sdks/python/apache_beam/transforms/core.py]: I thought we decided as we wanted to preserve the optimization for the sliding windows case where the sideinputs are globally windowed as that was a performance improvement.

I'm a bit confused about the API here so some help would be great.
Lets say we have 5 windows in the element and then:
case 1 (global side inputs): Process is called once with all windows together
case 2 (windowed side inputs): Process is called 5 times once for each window

What should happen here if we only expose one window at a time ?

Comment (robertwb) [inline sdks/python/apache_beam/transforms/core.py]: For the record, the fact that multiple windows are optimized in this way should be transparent to the user. If the window is exposed, or a side input depends on the window, we must call process multiple times.

**Comment by sb2nov** [comment]
Makes sense. I'll split the PR into multiple ones and then send them

**Comment by sb2nov** [comment]
Closing as this will be split into multple PRs
